homeassistant/components/lcn/__init__.py

```diff
@@ -150,6 +158,67 @@ async def async_unload_entry(
     return unload_ok
 
 
+def host_input_received(
+    hass: HomeAssistant,
+    config_entry: config_entries.ConfigEntry,
+    device_registry: dr.DeviceRegistry,
+    inp: pypck.inputs.Input,
+) -> None:
+    """Process received input object (command) from LCN bus."""
+    if not isinstance(inp, pypck.inputs.ModInput):
+        return
+
+    lcn_connection = hass.data[DOMAIN][config_entry.entry_id][CONNECTION]
+    logical_address = lcn_connection.physical_to_logical(inp.physical_source_addr)
+    address = (
+        logical_address.seg_id,
+        logical_address.addr_id,
+        logical_address.is_group,
+    )
+    identifiers = {(DOMAIN, generate_unique_id(config_entry.entry_id, address))}
+    device = device_registry.async_get_device(identifiers, set())
+
+    # fire event: lcn_transmitter, lcn_transponder or lcn_fingerprint
+    if isinstance(inp, pypck.inputs.ModStatusAccessControl):
+        event_data = {
+            "segment_id": address[0],
+            "module_id": address[1],
+            "code": inp.code,
+        }
+
+        if device is not None:
+            event_data.update({CONF_DEVICE_ID: device.id})
+
+        if inp.periphery == pypck.lcn_defs.AccessControlPeriphery.TRANSMITTER:
+            event_data.update(
+                {"level": inp.level, "key": inp.key, "action": inp.action.value}
+            )
+
+        event_name = f"lcn_{inp.periphery.value.lower()}"
+        hass.bus.async_fire(event_name, event_data)
+
+    # fire event: lcn_sendkeys
+    elif isinstance(inp, pypck.inputs.ModSendKeysHost):
+        for table, action in enumerate(inp.actions):
+            if action == pypck.lcn_defs.SendKeyCommand.DONTSEND:
+                continue
+
+            for key, selected in enumerate(inp.keys):
+                if not selected:
+                    continue
+                event_data = {
+                    "segment_id": address[0],
+                    "module_id": address[1],
+                    "key": pypck.lcn_defs.Key(table * 8 + key).name.lower(),
+                    "action": action.name.lower(),
+                }
+
+                if device is not None:
+                    event_data.update({CONF_DEVICE_ID: device.id})
+
+                hass.bus.async_fire("lcn_sendkeys", event_data)
```

Comment:
Maybe break this out into something like `_async_fire_send_keys_event(hass, device, address, inp)`